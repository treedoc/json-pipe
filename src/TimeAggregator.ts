import * as _ from 'lodash';

export class Statistic {
  total: number = 0;
  min: any;
  max: any;
  sum: number = 0;
  avg: number = 0;
  p50: number = 0;
  p90: number = 0;
  p99: number = 0;
  values: number[] = [];

  public addValue(val: number) {
    this.total++;
    this.values.push(val);
    if (this.min === undefined || val < this.min)
      this.min = val;   
    if (this.max === undefined || val > this.max)
      this.max = val;
    this.sum += val;
  }

  public calculate() {
    this.avg = this.sum / this.total;
    this.p50 = this.values[Math.floor(this.values.length * 0.5)];
    this.p90 = this.values[Math.floor(this.values.length * 0.9)];
    this.p99 = this.values[Math.floor(this.values.length * 0.99)];
    // this.values = [];
    return this;
  }
}

export class StatisticDataPoint  {
  readonly statistics: {[key: string]: Statistic} = {};
  constructor(public readonly time: number = 0) {
  }

  public calculate(): StatisticDataPoint {
    for (const key of Object.keys(this.statistics)) {
      this.statistics[key].calculate();
    }
    return this;
  }
}

export class TimeAggregator<T> {
  lastTime: number = 0;
  dataPoints: StatisticDataPoint[] = [];

  constructor(
    public timeField: string,
    public sampleInterval: number,
    // Number of sampler intervals, by default 1
    public timeWindow = 1,
    public classifier: string | ((data: T) => string),
    public dataExtractor: {[key: string]: string | ((data: T) => number)} = {},
    ) {
  }
  
  public onData(data: T): StatisticDataPoint[] {
    const result: StatisticDataPoint[] = [];
    if (!data) {  // End of stream
      this.dataPoints.forEach(d => d.calculate());
      return this.dataPoints;
    }
    
    const strTime = _.get(data, this.timeField);
    if (strTime === undefined || Number.isNaN(strTime)) {
      throw new Error(`Time field ${this.timeField} not found or not a number type in data ${JSON.stringify(data)}`);
    }
    const time = Number(strTime);

    const windowStart = Math.floor(time / this.sampleInterval) * this.sampleInterval;
    if (this.lastTime < windowStart) {
      const timeWindowStart = windowStart - this.sampleInterval * this.timeWindow;
      while(this.dataPoints.length > 0 && this.dataPoints[0].time <= timeWindowStart)
        result.push(this.dataPoints.shift()!.calculate());
      this.dataPoints.push(new StatisticDataPoint(windowStart));
    }

    const classifier = this.extractStringField(data, this.classifier);
    this.dataPoints.forEach(d => {
      const dataValues = this.extractNumberFields(data, this.dataExtractor);
      for (const key of Object.keys(dataValues)) {
        const dataKey = `${classifier}.${key}`;
        d.statistics[dataKey] = d.statistics[dataKey] || new Statistic();
        d.statistics[dataKey].addValue(dataValues[key]);
      }
    })
    
    this.lastTime = time;
    return result;
  }

  extractNumberField(data: T, field: string | ((data: T) => number)): number {
    if (typeof field === 'string') {
      return _.get(data, field);
    } else {
      return field(data);
    }
  }

  extractStringField(data: T, field: string | ((data: T) => string)): string {
    if (typeof field === 'string') {
      return _.get(data, field);
    } else {
      return field(data);
    }
  }

  extractNumberFields(data: T, field:{[key:string]: string | ((data: T) => number)}): {[key: string]: number} {
    const result: {[key: string]: number} = {};
    for (const key of Object.keys(field)) {
      result[key] = this.extractNumberField(data, field[key]);
    }
    return result;
  }
}
