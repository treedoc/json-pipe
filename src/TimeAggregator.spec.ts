// import { describe, expect, test } from "vitest";
import { TimeAggregator } from "./TimeAggregator";

const data = [
  { time: 1000, value: {a: 0, b: 7, name: "n1"} },
  { time: 1001, value: {a: 1, b: 6, name: "n2"} },
  { time: 1010, value: {a: 2, b: 5, name: "n1"} },
  { time: 1015, value: {a: 3, b: 4, name: "n1"} },
  { time: 1021, value: {a: 4, b: 3, name: "n2"} },
  { time: 1025, value: {a: 5, b: 2, name: "n1"} },
  { time: 1033, value: {a: 6, b: 1, name: "n2"} },
  { time: 1034, value: {a: 7, b: 0, name: "n2"} },
  null,
]

const data2 = [
  { time: 1000, value: {a: 0, name: "n1"} },
  { time: 1001, value: {a: 1, name: "n1"} },
  { time: 1010, value: {a: 2, name: "n1"} },
  { time: 1015, value: {a: 3, name: "n1"} },
  { time: 1021, value: {a: 4, name: "n1"} },
  { time: 1025, value: {a: 5, name: "n1"} },
  { time: 1033, value: {a: 6, name: "n1"} },
  { time: 1034, value: {a: 7, name: "n1"} },
  null,
]


describe('TimeAggregator', () => {
  const aggregator = new TimeAggregator("time", 10, 1, "value.name", {a: "value.a"});
  test('Simple', () => {
    const result: any[] = [];
    data.forEach(d => result.push(...aggregator.onData(d)));
    expect(result).toMatchSnapshot();
  });
});