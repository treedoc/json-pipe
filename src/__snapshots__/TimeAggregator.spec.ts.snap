// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeAggregator Simple 1`] = `
Array [
  StatisticDataPoint {
    "statistics": Object {
      "n1": Statistic {
        "avg": 1,
        "max": 1,
        "min": 1,
        "p50": 1,
        "p90": 1,
        "p99": 1,
        "sum": 1,
        "total": 1,
        "values": Array [],
      },
      "n1.a": Statistic {
        "avg": 0,
        "max": 0,
        "min": 0,
        "p50": 0,
        "p90": 0,
        "p99": 0,
        "sum": 0,
        "total": 1,
        "values": Array [],
      },
      "n1.b": Statistic {
        "avg": 7,
        "max": 7,
        "min": 7,
        "p50": 7,
        "p90": 7,
        "p99": 7,
        "sum": 7,
        "total": 1,
        "values": Array [],
      },
      "n2": Statistic {
        "avg": 1,
        "max": 1,
        "min": 1,
        "p50": 1,
        "p90": 1,
        "p99": 1,
        "sum": 1,
        "total": 1,
        "values": Array [],
      },
      "n2.a": Statistic {
        "avg": 1,
        "max": 1,
        "min": 1,
        "p50": 1,
        "p90": 1,
        "p99": 1,
        "sum": 1,
        "total": 1,
        "values": Array [],
      },
      "n2.b": Statistic {
        "avg": 6,
        "max": 6,
        "min": 6,
        "p50": 6,
        "p90": 6,
        "p99": 6,
        "sum": 6,
        "total": 1,
        "values": Array [],
      },
    },
    "time": 1000,
  },
  StatisticDataPoint {
    "statistics": Object {
      "n1": Statistic {
        "avg": 1,
        "max": 1,
        "min": 1,
        "p50": 1,
        "p90": 1,
        "p99": 1,
        "sum": 2,
        "total": 2,
        "values": Array [],
      },
      "n1.a": Statistic {
        "avg": 2.5,
        "max": 3,
        "min": 2,
        "p50": 3,
        "p90": 3,
        "p99": 3,
        "sum": 5,
        "total": 2,
        "values": Array [],
      },
      "n1.b": Statistic {
        "avg": 4.5,
        "max": 5,
        "min": 4,
        "p50": 4,
        "p90": 4,
        "p99": 4,
        "sum": 9,
        "total": 2,
        "values": Array [],
      },
    },
    "time": 1010,
  },
  StatisticDataPoint {
    "statistics": Object {
      "n1": Statistic {
        "avg": 1,
        "max": 1,
        "min": 1,
        "p50": 1,
        "p90": 1,
        "p99": 1,
        "sum": 1,
        "total": 1,
        "values": Array [],
      },
      "n1.a": Statistic {
        "avg": 5,
        "max": 5,
        "min": 5,
        "p50": 5,
        "p90": 5,
        "p99": 5,
        "sum": 5,
        "total": 1,
        "values": Array [],
      },
      "n1.b": Statistic {
        "avg": 2,
        "max": 2,
        "min": 2,
        "p50": 2,
        "p90": 2,
        "p99": 2,
        "sum": 2,
        "total": 1,
        "values": Array [],
      },
      "n2": Statistic {
        "avg": 1,
        "max": 1,
        "min": 1,
        "p50": 1,
        "p90": 1,
        "p99": 1,
        "sum": 1,
        "total": 1,
        "values": Array [],
      },
      "n2.a": Statistic {
        "avg": 4,
        "max": 4,
        "min": 4,
        "p50": 4,
        "p90": 4,
        "p99": 4,
        "sum": 4,
        "total": 1,
        "values": Array [],
      },
      "n2.b": Statistic {
        "avg": 3,
        "max": 3,
        "min": 3,
        "p50": 3,
        "p90": 3,
        "p99": 3,
        "sum": 3,
        "total": 1,
        "values": Array [],
      },
    },
    "time": 1020,
  },
  StatisticDataPoint {
    "statistics": Object {
      "n2": Statistic {
        "avg": 1,
        "max": 1,
        "min": 1,
        "p50": 1,
        "p90": 1,
        "p99": 1,
        "sum": 2,
        "total": 2,
        "values": Array [],
      },
      "n2.a": Statistic {
        "avg": 6.5,
        "max": 7,
        "min": 6,
        "p50": 7,
        "p90": 7,
        "p99": 7,
        "sum": 13,
        "total": 2,
        "values": Array [],
      },
      "n2.b": Statistic {
        "avg": 0.5,
        "max": 1,
        "min": 0,
        "p50": 0,
        "p90": 0,
        "p99": 0,
        "sum": 1,
        "total": 2,
        "values": Array [],
      },
    },
    "time": 1030,
  },
]
`;
